ocaml backend fixes for pc, export offline lifter library

ailrst opened the pull request:
- generated dune file now has public name `asli.offline` so we can link it in gtirb_semantics
- ocaml offline lifter takes argument `?pc`, with the offline lifter producing a run-time error if its required and not provided. If the generated lifter is not configured to supply pc it just ignores the argument.

ailrst committed:
tweak ocaml backend support pc, and export a library

ailrst committed:
update stub lifter

katrinafyi commented on libASL/ocaml_backend.ml:
```diff
@@ -65,8 +66,15 @@ let write_preamble opens st =
-  Printf.fprintf st.oc "let run enc =\n  reset_ir ();\n  %s enc;\n  get_ir ()\n" (name_of_ident fid)
+let write_epilogue use_pc fid st =
+  let conv_pc = "let pc = (mkBits (Z.of_int 64) (Z.of_int (Option.get pc))) in" in
+  let dis_call = (match use_pc with 
```
just one comment, i feel like it's redundant to make the generated ocaml function have a `pc : int option` parameter if it fails when pc is None. you can just include or exclude an int parameter, based on the gen-time use_pc variable.

otherwise, l gee tm

ailrst replied:
This is an optional parameter so that the api is the same with and without pc support, and you get a runtime error rather than a compile time error. Failing at compile time is more annoying because it prevents you from running tests etc. It should probably have two different package names with different apis instead, but you'd want to be able to build both at the same time.

ailrst committed:
new offline lifter & asli packages, make dune build offline lifter

We now define two packages, asli and aslp_offline. The offline
lifter library gets a different name depending on whether
its built with pc support which is enabled by default in
offlineASL/gen-command (breaking offline-coverage).

libASL/dune now invokes asli to generate the lifter into the
build directory. This makes everything buildable
so the offline lifter should be useable as an opam pin dependency.

Asli can generate the dune to build the offline lifter,
but there is a little bit bit of manual mess, e.g. regarding
the command to build asli. We also don't properly invalidate the
cached marshalled generated lifter (which appears in
_build/default/marshalled-offline-lifter-*) when we rebuild
asli; to regenerate the lifter you must first run `dune clean`.

ailrst commented:
The last commit separates offline and asli into two separate packages, and makes it possible to build the offline lifter directly with dune (see commit message). If installing through opam asli needs to be explicitly installed first otherwise it tries to find it in the main opam repository.  

```
$ opam pin scan 'git+https://github.com/UQ-PAC/aslp.git#export-offline-lifter'
# Name        # Version  # Url                                                         # Subpath
aslp_offline  0.2.0      git+https://github.com/UQ-PAC/aslp.git#export-offline-lifter  -
asli          0.2.0      git+https://github.com/UQ-PAC/aslp.git#export-offline-lifter  -
```

katrinafyi commented:
should we turn off marshalling by default

katrinafyi committed:
make `dune build` build just asli

the install dependency was requiring a build of all packages.
note, to install asli only, you will have to use `dune install -p asli`

katrinafyi committed:
make `dune build offlineASL` build all aslp offline things

this will also build asli if needed for the offline generation

katrinafyi committed:
readme: recommend dune to install package, clarify subpackages

ailrst committed:
separate offline lifter with pc support

ailrst committed:
offline depend on stage0

- Moves Primops, Value and Symbolic into stage0 to support this
- Some fixes for the previous commit

katrinafyi commented:
i saw this error while building via an opam pin (using the instructions at https://www.github.com/katrinafyi/aslp-web/pull/8). idk if this would happen with dune in a local checkout.

```ocaml
#=== ERROR while compiling aslp_offline.0.2.0 =================================#
# context     2.3.0 | linux/x86_64 | ocaml.4.14.2 | pinned(git+https://github.com/UQ-PAC/aslp#export-offline-lifter#a870076de9b6f998120db2e1c910539338c7315f)
# path        ~/.opam/default/.opam-switch/build/aslp_offline.0.2.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p aslp_offline -j 7 @install
# exit-code   1
# env-file    ~/.opam/log/aslp_offline-114121-dcc48d.env
# output-file ~/.opam/log/aslp_offline-114121-dcc48d.out
### output ###
# Failed to gen 'aarch64_integer_pac_strip_dp_1src.0([],[LiftTime])': Unexpected runtime expression: gen_add_int.0 {{  }} ( gen_add_int.0 {{  }} ( gen_sub_int.0 {{  }} ( gen_sub_int.0 {{  }} ( gen_int_lit.0 {{  }} ( 56 ),gen_cvt_bits_uint.0 {{ 8 }} ( gen_load.0 {{  }} ( CalculateBottomPACBit143__3 ) ) ),gen_int_lit.0 {{  }} ( 1 ) ),gen_int_lit.0 {{  }} ( 1 ) ),gen_int_lit.0 {{  }} ( 8 ) )
# [...]
# Done generating ocaml lifter into '.'.
# (cd _build/default && /home/rina/.opam/default/bin/ocamlc.opt -cclib -lstdc++ -g -bin-annot -I bin/.offline_sem.eobjs/byte -I /home/rina/.opam/default/lib/asli/libASL -I /home/rina/.opam/default/lib/asli/libASL-stage0 -I /home/rina/.opam/default/lib/asli/libASL-stage1 -I /home/rina/.opam/default/lib/asli/libASL-support-native -I /home/rina/.opam/default/lib/asli/libASL-virtual -I /home/rina/.[...]
# File "bin/offline_sem.ml", line 2, characters 5-12:
# 2 | open Asl_ast
#          ^^^^^^^
# Error: Unbound module Asl_ast
# (cd _build/default && /home/rina/.opam/default/bin/ocamlc.opt -cclib -lstdc++ -g -bin-annot -I bin/.offline_coverage.eobjs/byte -I /home/rina/.opam/default/lib/asli/libASL -I /home/rina/.opam/default/lib/asli/libASL-stage0 -I /home/rina/.opam/default/lib/asli/libASL-stage1 -I /home/rina/.opam/default/lib/asli/libASL-support-native -I /home/rina/.opam/default/lib/asli/libASL-virtual -I /home/r[...]
# File "bin/offline_coverage.ml", line 3, characters 5-12:
# 3 | open Asl_ast
#          ^^^^^^^
# Error: Unbound module Asl_ast
```

ailrst commented:
> i saw this error while building via an opam pin (using the instructions at https://www.github.com/katrinafyi/aslp-web/pull/8). idk if this would happen with dune in a local checkout.
> 
> ```ocaml
> #=== ERROR while compiling aslp_offline.0.2.0 
>...
> # Error: Unbound module Asl_ast
> ```

I can't reproduce this on either of my computers, here with tool versions:

```
$ ocaml --version ; dune --version ; opam --version 
The OCaml toplevel, version 5.2.1
3.17.1
2.2.1
```

I don't know why it would do that as offline_sem is still linking libASL directly.

katrinafyi commented:
it looks like it was a my computer problem, my /home/rina/.opam/default/lib/asli/ was an scattered mix of multiple asli versions

ailrst committed:
update offline dune files

ailrst committed:
fix bin/dune

ailrst committed:
fix test action coverage

ailrst committed:
add js lib

katrinafyi committed:
ci: try to fake DUNE_INSTALL_PREFIX

katrinafyi committed:
readme: add alias for building all offline things

katrinafyi committed:
bootstrapping for offline dune files

katrinafyi committed:
tests: add package restrictions

ailrst committed:
reword readme

ailrst commented:
Your changes look really good, `dune build @offline-coverage` works now from a fresh repo, although it unneccessarily builds the offline lifter with lift-time pc support, that isn't too much of an issue.
